DispatcherTest.testForwardToWelcome() fails in ee9/ee8 if surefire reuseForks=true

**Jetty version(s)**
12.0.10

**Jetty Environment**
ee9 / ee8

**Java version/vendor** `(use: java -version)`
Any

**OS type/version**
Any

**Description**
When working on PR #11930 it was discovered that the ee9 / ee8 (not ee10) DispatcherTests are failing in the following way ...

*Stacktrace:*

```
java.lang.AssertionError: 
Expected: a string containing "HTTP/1.1 302 Found"
     but: was "HTTP/1.1 200 OK
Last-Modified: Thu, 20 Jun 2024 21:16:47 GMT
Accept-Ranges: bytes
Content-Length: 4096
Connection: close
index.html
"
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
	at org.eclipse.jetty.ee9.servlet@12.0.11-SNAPSHOT/org.eclipse.jetty.ee9.servlet.DispatcherTest.testForwardToWelcome(DispatcherTest.java:159)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
```

*Standard Error*

```
Running org.eclipse.jetty.ee9.servlet.DispatcherTest.testForwardToWelcome()
2024-06-20 21:45:04.094:INFO :oejs.Server:ForkJoinPool-1-worker-3: jetty-12.0.11-SNAPSHOT; built: 2024-06-20T21:17:46.344Z; git: ce6174f9ff9dba9fcc68859880d17c9a16a1401e; jvm 21.0.1+12-LTS
2024-06-20 21:45:04.095:INFO :oejsh.ContextHandler:ForkJoinPool-1-worker-3: Started oeje9n.ContextHandler$CoreContextHandler@661cd22a{/context,/context,b=(URL) file:///home/jenkins/agent/workspace/jetty.project_PR-11930/jetty-ee9/jetty-ee9-servlet/src/test/resources/contextResources/,a=AVAILABLE,h=oeje9n.ContextHandler$CoreContextHandler$CoreToNestedHandler@64c64de3{STARTED}}
2024-06-20 21:45:04.095:INFO :oejsh.ContextHandler:ForkJoinPool-1-worker-3: Started oeje9n.ContextHandler$CoreContextHandler@c8de777{/resource,/resource,b=null,a=AVAILABLE,h=oeje9n.ContextHandler$CoreContextHandler$CoreToNestedHandler@5af20a50{STARTED}}
2024-06-20 21:45:04.095:INFO :oejs.AbstractConnector:ForkJoinPool-1-worker-3: Started LocalConnector@71e78163{HTTP/1.1, (http/1.1)}
2024-06-20 21:45:04.095:INFO :oejs.Server:ForkJoinPool-1-worker-3: Started oejs.Server@3c129861{STARTING}[12.0.11-SNAPSHOT,sto=0] @7303ms
2024-06-20 21:45:04.099:INFO :oejs.Server:ForkJoinPool-1-worker-3: Stopped oejs.Server@3c129861{STOPPING}[12.0.11-SNAPSHOT,sto=0]
2024-06-20 21:45:04.099:INFO :oejs.AbstractConnector:ForkJoinPool-1-worker-3: Stopped LocalConnector@71e78163{HTTP/1.1, (http/1.1)}
```

That PR disabled the surefire reuseForks to get past this issue.

https://github.com/jetty/jetty.project/blob/f78f4422f5913820bead90cfa7a337415ff8935d/jetty-ee9/jetty-ee9-servlet/pom.xml#L73-L76

Turns out there is some unintended interaction between test cases when resuseForks = true.

To replicate, just set `<reuseForks>true</reuseForks>` in the surefire config on project `/ee9/jetty-ee9-servlet/` and then run `mvn clean install -pl :jetty-ee9-servlet` to build & run all tests.
It will fail, reliably, with the above stacktrace.

But if you run just the `DispatcherTest` it works reliably 100% of the time.

In talking with @olamy the current theory is that there is some kind of static field / class / instance that is hanging around between runs on the same fork that is impacting the execution of `DispatcherTest` in particular.

--- a/jetty-ee9/jetty-ee9-servlet/pom.xml
+++ b/jetty-ee9/jetty-ee9-servlet/pom.xml
@@ -73,7 +73,7 @@
         <artifactId>maven-surefire-plugin</artifactId>
         <configuration>
           <!-- Reuse Forks causes test failures -->
-          <reuseForks>false</reuseForks>
+          <reuseForks>true</reuseForks>
           <argLine>@{argLine} ${jetty.surefire.argLine}
             --add-modules org.eclipse.jetty.util.ajax
             --add-reads org.eclipse.jetty.ee9.servlet=org.eclipse.jetty.logging</argLine>

--- a/jetty-ee9/jetty-ee9-servlet/src/test/java/org/eclipse/jetty/ee9/servlet/DefaultServletTest.java
+++ b/jetty-ee9/jetty-ee9-servlet/src/test/java/org/eclipse/jetty/ee9/servlet/DefaultServletTest.java
@@ -26,6 +26,7 @@ import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.EnumSet;
 import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
@@ -2569,44 +2570,52 @@ public class DefaultServletTest
     @Test
     public void testGetPrecompressedSuffixMapping() throws Exception
     {
-        Path docRoot = workDir.getEmptyPathDir().resolve("docroot");
-        FS.ensureDirExists(docRoot);
-
-        startServer((context) ->
+        final AtomicReference<ResourceFactory> oldFileResourceFactory = new AtomicReference<>();
+        try
         {
-            ResourceFactory.registerResourceFactory("file", new URLResourceFactory());
-            Resource resource = ResourceFactory.of(context).newResource(docRoot);
-            assertThat("Expecting URLResource", resource.getClass().getName(), endsWith("URLResource"));
-            context.setBaseResource(resource);
+            Path docRoot = workDir.getEmptyPathDir().resolve("docroot");
+            FS.ensureDirExists(docRoot);
 
-            ServletHolder defholder = context.addServlet(DefaultServlet.class, "*.js");
-            defholder.setInitParameter("cacheControl", "no-store");
-            defholder.setInitParameter("dirAllowed", "false");
-            defholder.setInitParameter("gzip", "false");
-            defholder.setInitParameter("precompressed", "gzip=.gz");
-        });
+            startServer((context) ->
+            {
+                oldFileResourceFactory.set(ResourceFactory.unregisterResourceFactory("file"));
+                ResourceFactory.registerResourceFactory("file", new URLResourceFactory());
+                Resource resource = ResourceFactory.of(context).newResource(docRoot);
+                assertThat("Expecting URLResource", resource.getClass().getName(), endsWith("URLResource"));
+                context.setBaseResource(resource);
 
+                ServletHolder defholder = context.addServlet(DefaultServlet.class, "*.js");
+                defholder.setInitParameter("cacheControl", "no-store");
+                defholder.setInitParameter("dirAllowed", "false");
+                defholder.setInitParameter("gzip", "false");
+                defholder.setInitParameter("precompressed", "gzip=.gz");
+            });
 
-        FS.ensureDirExists(docRoot.resolve("scripts"));
+            FS.ensureDirExists(docRoot.resolve("scripts"));
 
-        String scriptText = "This is a script";
-        Files.writeString(docRoot.resolve("scripts/script.js"), scriptText, UTF_8);
+            String scriptText = "This is a script";
+            Files.writeString(docRoot.resolve("scripts/script.js"), scriptText, UTF_8);
 
-        byte[] compressedBytes = compressGzip(scriptText);
-        Files.write(docRoot.resolve("scripts/script.js.gz"), compressedBytes);
+            byte[] compressedBytes = compressGzip(scriptText);
+            Files.write(docRoot.resolve("scripts/script.js.gz"), compressedBytes);
 
-        String rawResponse = connector.getResponse("""
-            GET /context/scripts/script.js HTTP/1.1
-            Host: test
-            Accept-Encoding: gzip
-            Connection: close
-            
-            """);
-        HttpTester.Response response = HttpTester.parseResponse(rawResponse);
-        assertThat(response.getStatus(), is(HttpStatus.OK_200));
-        assertThat("Suffix url-pattern mapping not used", response.get(HttpHeader.CACHE_CONTROL), is("no-store"));
-        String responseDecompressed = decompressGzip(response.getContentBytes());
-        assertThat(responseDecompressed, is("This is a script"));
+            String rawResponse = connector.getResponse("""
+                GET /context/scripts/script.js HTTP/1.1
+                Host: test
+                Accept-Encoding: gzip
+                Connection: close
+                            
+                """);
+            HttpTester.Response response = HttpTester.parseResponse(rawResponse);
+            assertThat(response.getStatus(), is(HttpStatus.OK_200));
+            assertThat("Suffix url-pattern mapping not used", response.get(HttpHeader.CACHE_CONTROL), is("no-store"));
+            String responseDecompressed = decompressGzip(response.getContentBytes());
+            assertThat(responseDecompressed, is("This is a script"));
+        }
+        finally
+        {
+            ResourceFactory.registerResourceFactory("file", oldFileResourceFactory.get());
+        }
     }
 
     @Test

